src/Microsoft.DotNet.CodeFormatting/Rules/AttributeSeparateListsRule.cs

```diff
@@ -0,0 +1,84 @@
+        {
+            public override SyntaxNode VisitParameter(ParameterSyntax node)
+            {
+                // We don't want to flatten the attribute lists for parameters. Those are
+                // usually short, such as [In, Out] and collapsing them can actually
+                // improve readability.
+                return node;
+            }
+
+            public override SyntaxList<TNode> VisitList<TNode>(SyntaxList<TNode> list)
+            {
+                list = base.VisitList(list);
+
+                if (typeof (TNode) != typeof (AttributeListSyntax))
+                    return list;
+
+                var attributeLists = (SyntaxList<AttributeListSyntax>) (object) list;
+                return (SyntaxList<TNode>) (object) VisitAttributeLists(attributeLists);
+            }
+
+            private static SyntaxList<AttributeListSyntax> VisitAttributeLists(SyntaxList<AttributeListSyntax> attributeLists)
+            {
+                var result = new List<AttributeListSyntax>();
+
+                foreach (var attributeList in attributeLists)
+                {
+                    var firstIndex = result.Count;
+
+                    for (var i = 0; i < attributeList.Attributes.Count; i++)
+                    {
+                        var attribute = attributeList.Attributes[i];
+                        var separatorTrivia = i < attributeList.Attributes.Count - 1
+                                                ? attributeList.Attributes.GetSeparator(i).GetAllTrivia()
+                                                : Enumerable.Empty<SyntaxTrivia>();
+
+                        var attributeWithoutTrivia = attribute.WithoutLeadingTrivia().WithoutTrailingTrivia();
+                        var singletonList = SyntaxFactory.AttributeList(attributeList.Target, SyntaxFactory.SeparatedList(new[] { attributeWithoutTrivia }))
+                                                         .WithLeadingTrivia(attribute.GetLeadingTrivia())
+                                                         .WithTrailingTrivia(attribute.GetTrailingTrivia().Concat(separatorTrivia));
+                        result.Add(singletonList);
+                    }
+
+                    var lastIndex = result.Count - 1;
+
+                    var leadingTrivia = attributeList.GetLeadingTrivia()
+                                            .Concat(attributeList.OpenBracketToken.TrailingTrivia)
+                                            .Concat(result[firstIndex].GetLeadingTrivia());
+
+                    var trailingTrivia = result[lastIndex].GetTrailingTrivia()
+                                            .Concat(attributeList.CloseBracketToken.LeadingTrivia)
+                                            .Concat(attributeList.GetTrailingTrivia());
+
+                    result[firstIndex] = result[firstIndex].WithLeadingTrivia(leadingTrivia);
+                    result[lastIndex] = result[lastIndex].WithTrailingTrivia(trailingTrivia);
```

Comment:
I've been staring at this for ~5 minutes now thinking about the trivia implications.  Consider the following:

``` csharp
[Attr1,
    Attr2]
```

I'm fairly certain this code would end up retaining the extra new line and spacing with the above logic.  

I'm struggling to think of a better approach though.  I considered recommending that you strip everything but comments and structured trivia and add that back in but that has some pathological cases to consider.  

Trivia isn't trivial :(